app/services/cart_service.rb

```diff
@@ -15,28 +15,35 @@ def populate(from_hash, overwrite = false)
     @distributor, @order_cycle = distributor_and_order_cycle
 
     @order.with_lock do
-      variants = read_variants from_hash
-      attempt_cart_add_variants variants
-      overwrite_variants variants unless !overwrite
+      variants_data = read_variants from_hash
+      attempt_cart_add_variants variants_data
+      overwrite_variants variants_data if overwrite
     end
     valid?
   end
 
-  def attempt_cart_add_variants(variants)
-    variants.each do |v|
-      if varies_from_cart(v)
-        attempt_cart_add(v[:variant_id], v[:quantity], v[:max_quantity])
-      end
+  private
+
+  def attempt_cart_add_variants(variants_data)
+    loaded_variants = indexed_variants(variants_data)
```

Comment:
Here we're loading all the variants we will need in the following code in advance (with their associated objects eager-loaded), to avoid loading them all one-by-one (and loading all their associated objects one-by-one as well). 

I did some renaming of variables in this class to distinguish between the hash of `variant_data` that's coming from posted `params`, and actual variant objects.